service/frontend/workflow_handler.go

```diff
@@ -3900,6 +3952,59 @@ func (wh *WorkflowHandler) getArchivedHistory(
 	}, nil
 }
 
+func (wh *WorkflowHandler) getArchivedHistoryReverse(
+	ctx context.Context,
+	request *workflowservice.GetWorkflowExecutionHistoryReverseRequest,
+	namespaceID namespace.ID,
+) (*workflowservice.GetWorkflowExecutionHistoryReverseResponse, error) {
+	entry, err := wh.namespaceRegistry.GetNamespaceByID(namespaceID)
+	if err != nil {
+		return nil, err
+	}
+
+	URIString := entry.HistoryArchivalState().URI
+	if URIString == "" {
+		// if URI is empty, it means the namespace has never enabled for archival.
+		// the error is not "workflow has passed retention period", because
+		// we have no way to tell if the requested workflow exists or not.
+		return nil, errHistoryNotFound
+	}
+
+	URI, err := archiver.NewURI(URIString)
+	if err != nil {
+		return nil, err
+	}
+
+	historyArchiver, err := wh.archiverProvider.GetHistoryArchiver(URI.Scheme(), string(primitives.FrontendService))
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := historyArchiver.Get(ctx, URI, &archiver.GetHistoryRequest{
+		NamespaceID:   namespaceID.String(),
+		WorkflowID:    request.GetExecution().GetWorkflowId(),
+		RunID:         request.GetExecution().GetRunId(),
+		NextPageToken: request.GetNextPageToken(),
+		PageSize:      int(request.GetMaximumPageSize()),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	history := &historypb.History{}
+	for _, batch := range resp.HistoryBatches {
+		history.Events = append(history.Events, batch.Events...)
+	}
+	// reverse the events
```

Comment:
I am a bit confused here. Looks like this is only reversing events within a page. When there are multiple pages, the returned history will look like something like:
event 100, 99, 98, ..., 1, event 200, 199, 198, ..., 101.